doc/specs/#5000 - Process Model 2.0/#5000 - Process Model 2.0.md

```diff
@@ -0,0 +1,1253 @@
+---
+author: Mike Griese @zadjii-msft
+created on: 2020-07-31
+last updated: 2020-10-30
+issue id: #5000
+---
+
+# Windows Terminal Process Model 2.0
+
+## Abstract
+
+The Windows Terminal currently exists as a single process per window, with one
+connection per terminal pane (which could be an additional conpty process and
+associated client processes). This model has proven effective for the simple
+windowing we've done so far. However, in order to support scenarios like
+dragging tabs into other windows, or having one top-level window with different
+elevation levels within it, this single process model will not be sufficient.
+
+This spec outlines changes to the Terminal process model in order to enable the
+following scenarios:
+
+* Tab Tearoff/ Reattach ([#1256])
+* Run `wt` in the current window ([#4472])
+* Single Instance Mode ([#2227])
+* Quake Mode ([#653])
+* Mixed Elevation ([#1032] & [#632])
+
+## Inspiration
+
+Much of the design for this feature was inspired by (what I believe is) the web
+browser process model. For a web browser, there's often a separate process for

```

Comment:
Technically this is the old way. Modern browsers, for example, Firefox will attempt to use no more than 8 processes (based on the core count of the machine) to host all the tabs, in order to limit the process overhead when people opens large amount of tabs, like 100 tabs in a window.

Reply:
I think people don't really need that much tabs inside a terminal. But if they do, we'd better be prepared for it in the 2.0 process model.

Reply:
Notes added, thanks!